In [1]:
import pandas as pd
import numpy as np


In [21]:
titanic = pd.read_csv("Titanic-Dataset.csv")
titanic

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [22]:
print("Number of rows:", titanic.shape[0])
print("Number of columns:", titanic.shape[1])


Number of rows: 891
Number of columns: 12


In [4]:
display(titanic.head())


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [5]:
display(titanic.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


None

In [7]:
display(titanic.describe())

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [8]:
print(titanic.isnull().sum())

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


In [11]:
titanic['Age'] = titanic['Age'].fillna(titanic['Age'].median())
titanic['Cabin'] = titanic['Cabin'].fillna("Unknown")
titanic['Embarked'] = titanic['Embarked'].fillna(titanic['Embarked'].mode()[0])
print(titanic.isnull().sum())


PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64


In [12]:
print(len(titanic))

891


In [15]:
titanic = titanic.drop_duplicates()
print(len(titanic))

891


In [26]:
titanic['Age'] = titanic['Age'].astype(float)
titanic['Fare'] = titanic['Fare'].astype(float)

In [25]:
titanic = titanic.rename(columns={'Sex': 'Gender', 'Pclass': 'Passenger_Class'})

In [27]:
titanic['Gender'] = titanic['Gender'].map({'male': 0, 'female': 1})
titanic['Embarked'] = titanic['Embarked'].map({'S': 0, 'C': 1, 'Q': 2})

titanic[['Age', 'Fare']] = titanic[['Age', 'Fare']].apply(lambda x: (x - x.min()) / (x.max() - x.min()))

titanic['Family_Size'] = titanic.apply(lambda row: row['SibSp'] + row['Parch'] + 1, axis=1)
print(titanic)

     PassengerId  Survived  Passenger_Class  \
0              1         0                3   
1              2         1                1   
2              3         1                3   
3              4         1                1   
4              5         0                3   
..           ...       ...              ...   
886          887         0                2   
887          888         1                1   
888          889         0                3   
889          890         1                1   
890          891         0                3   

                                                  Name  Gender       Age  \
0                              Braund, Mr. Owen Harris       0  0.271174   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...       1  0.472229   
2                               Heikkinen, Miss. Laina       1  0.321438   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)       1  0.434531   
4                             Allen, Mr. William Henry 

In [28]:
# Shape of dataset after all the cleaning
print("Number of rows:", titanic.shape[0])
print("Number of columns:", titanic.shape[1])


Number of rows: 891
Number of columns: 13


                                    SUMMARY
I cleaned and preprocessed Titanic dataset cleaned and preprocessed using Pandas and NumPy
1. Data Loading and Exploration

Loaded dataset using pd.read_csv()

Checked basic info with info() and describe()

Displayed shape → 891 rows × 12 columns

2. Handling Missing Values

Detected missing values using isnull().sum()

Filled:

Age with median

Cabin with "Unknown"

Embarked with most frequent value (mode)

Verified no missing values remained

3. Removing Duplicates

Checked duplicates with duplicated().sum()

Dropped duplicates using drop_duplicates()

None were found

4. Data Type Conversion

Converted Age and Fare to float for numerical consistency

5. Renaming Columns

Sex → Gender

Pclass → Passenger_Class

6. Encoding Categorical Variables

Used map() to encode:

Gender: male = 0, female = 1

Embarked: S = 0, C = 1, Q = 2

7. Feature Engineering

Added Family_Size = SibSp + Parch + 1

8. Scaling Numerical Features

Normalized Age and Fare using Min–Max scaling with apply()

9. Final Output

No missing or duplicate values

12 columns total, including new engineered feature

Ready for data analysis or machine learning

